.github/workflows/test_on_push.yml

```diff
@@ -59,36 +54,52 @@ jobs:
     - name: Install standard python dependencies
       run: |
         python -m pip install --upgrade pip wheel setuptools
-        python -m pip install tox
-
+        python -m pip install nox
+    
     - name: Run unit tests (except GNU/Linux with Python 3.9)
       if: matrix.os != 'ubuntu-latest' || matrix.python-version != 3.9
-      run: python -m tox -e unit
+      run: nox -p "${{matrix.python-version}}" -s unit
 
-    - name: Run unit tests for GNU/Linux with Python 3.9 and generate coverage report
-      if: matrix.os == 'ubuntu-latest' && matrix.python-version == 3.9
-      run: tox -e coverage
+    - name: Run integration tests
+      run: nox -p "${{matrix.python-version}}" -s integration
 
-    - name: Upload coverage report
+    - name: Install docs dependencies and run doctests
       if: matrix.os == 'ubuntu-latest' && matrix.python-version == 3.9
-      uses: codecov/codecov-action@v2.1.0
+      run: nox -s sphinx
 
-    - name: Run integration tests
-      run: python -m tox -e integration
+    - name: Install dev dependencies and run example tests
+      if: matrix.os == 'ubuntu-latest'
+      run: nox -p "${{matrix.python-version}}" -s examples
```

Comment:
The examples take a bit long to run, could we move the coverage before the examples so we get the report without having to wait for the examples to be done?

In PyBaMM we only run the coverage on one OS and version (e.g Ubuntu with Python 3.11) and that's done instead of the unit tests (as the coverage already runs the unit tests, but a bit slower).